### PR TITLE
Upgrade node, rush, and pnpm to latest

### LINKED ISSUE #681: Upgrade node, rush, and pnpm to latest
Current versions not compatible with Node 20

### PR DESCRIPTION
**Pull request by mikeharder**
* Fixes #681 
* Upgrade minimum Node version to 18
  * package.json/engines
  * rush.json
* Upgrade default Node version to LTS / 20
  * CONTRIBUTING.md
  * pipelines
* Upgrade rush from 5.62.1 to 5.121.0
  * Update generated "common/scripts" folder
* Upgrade pnpm from 6.22.1 to 8.15.7
  * Upgrade pnpm-lock.yaml from v5 to v6
  * pnpm@9 released last week, but too soon to upgrade
* Includes "rush update --full"
  * Recommended to upgrade all deps when upgrading node/rush/pnpm 
* In tests, sort results before comparing to ensure stable order

## Open Questions
* Should we publish a release of these packages with this upgrade?  Or just consider this an infrastructure upgrade, and wait until the next time someone wants to publish the packages for another reason?

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -60,21 +60,13 @@ For more information see the [Code of Conduct FAQ](https://opensource.microsoft.
...
-   ```bash
-   # Install latest rush, globally
-   npm install -g @microsoft/rush
-
-   # OR Install specific version, to the local node_modules folder
-   cd "this_repo_local_clone_dir"
-   npm install @microsoft/rush@5.26.1
-   ```
+5. Install [@Microsoft/Rush](https://rushjs.io/).
+   - `npm install -g @microsoft/rush`
```
Comment (mikeharder) [inline CONTRIBUTING.md:69]: `rush` is a bootstrapper with it's own internal versioning mechanism, so it's one of the rare packages I recommend installing latest globally.  For the convenience of just typing `rush <command>` in your inner loop.

In CI, or if you didn't want install it globally for inner loop, you should call the script in the repo which automatically does a local install and runs the correct version:

```
$ node common/scripts/install-run-rush.js <arguments>
```

File: rush.json [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
-  "rushVersion": "5.62.1",
-  "pnpmVersion": "6.22.1",
+  "rushVersion": "5.121.0",
+  "pnpmVersion": "8.15.7",
   "ensureConsistentVersions": true,
   "pnpmOptions": {
     "strictPeerDependencies": true
   },
-  "nodeSupportedVersionRange": ">=12.0.0",
+  "nodeSupportedVersionRange": ">=18.0.0",
```
Comment (mikeharder) [inline rush.json:9]: We should support lowest LTS (currently 18)

File: rush.json [hunk 3]
```diff
@@ -1,12 +1,12 @@
 {
   "$schema": "https://developer.microsoft.com/json-schemas/rush/v5/rush.schema.json",
-  "rushVersion": "5.62.1",
-  "pnpmVersion": "6.22.1",
+  "rushVersion": "5.121.0",
+  "pnpmVersion": "8.15.7",
```
Comment (mikeharder) [inline rush.json:4]: pnpm@9 is technically "latest", but it was just released last week so I'd rather wait a few months for it to bake.  pnpm@8 will be maintained for years to come.

File: packages/rulesets/package.json [hunk 4]
```diff
@@ -14,7 +14,7 @@
     "lint": "eslint ./src --ext .ts --max-warnings=0"
   },
   "engines": {
-    "node": ">=12"
+    "node": ">=18"
```
Comment (mikeharder) [inline packages/rulesets/package.json:17]: We should support lowest LTS (currently 18)

File: packages/rulesets/package.json [hunk 5]
```diff
@@ -71,7 +71,7 @@
   },
   "dependencies": {
     "@apidevtools/json-schema-ref-parser": "^9.0.9",
-    "@stoplight/json-ref-resolver": "3.1.3",
+    "@stoplight/json-ref-resolver": "^3.1.6",
```
Comment (mikeharder) [inline packages/rulesets/package.json:74]: - Failures when `3.1.3` and `3.1.6` were loaded side-by-side
- Unclear from git history why this dep was pinned when related deps floated

File: packages/azure-openapi-validator/autorest/package.json [hunk 6]
```diff
@@ -22,7 +22,7 @@
     "validator"
   ],
   "engines": {
-    "node": ">=12"
+    "node": ">=18"
```
Comment (mikeharder) [inline packages/azure-openapi-validator/autorest/package.json:25]: We should support lowest LTS (currently 18)

File: packages/azure-openapi-validator/autorest/package.json [hunk 7]
```diff
@@ -56,7 +56,7 @@
     "@stoplight/spectral-ruleset-bundler": "^1.2.1",
     "@stoplight/spectral-rulesets": "~1.6.0",
     "@stoplight/spectral-runtime": "^1.1.0",
-    "@stoplight/json-ref-resolver": "3.1.3",
+    "@stoplight/json-ref-resolver": "^3.1.6",
```
Comment (mikeharder) [inline packages/azure-openapi-validator/autorest/package.json:59]: - Failures when `3.1.3` and `3.1.6` were loaded side-by-side
- Unclear from git history why this dep was pinned when related deps floated

File: azure-pipelines/common-build.yml [hunk 8]
```diff
@@ -1,6 +1,6 @@
 parameters:
 - name: 'NodeVersion'
-  default: '12.x'
+  default: '20.x'
```
Comment (mikeharder) [inline azure-pipelines/common-build.yml:3]: Latest LTS

File: azure-pipelines/staging-release-pipelines.yml [hunk 9]
```diff
@@ -20,8 +20,6 @@ extends:
 
         steps:
           - template: common-build.yml
-            parameters: 
-              NodeVersion: "16.x"
```
Comment (mikeharder) [inline azure-pipelines/staging-release-pipelines.yml:23]: Template defaults to latest LTS, so override no longer needed

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -60,21 +60,13 @@ For more information see the [Code of Conduct FAQ](https://opensource.microsoft.
 
 1. Install [nvm] ([nvm for Windows]), a tool to manage `Node.js` versions.
 2. Restart your terminal so that `nvm` command is recognized.
-3. Using `nvm`, install `Node.js` version `12.22.12`. This is the version used by the [PR CI pipeline] to build `LintDiff`.
-   - `nvm install 12.22.12`
-   - `nvm use 12.22.12`
+3. Using `nvm`, install `Node.js` version `LTS`. This is the version used by the [PR CI pipeline] to build `LintDiff`.
+   - Linux: `nvm install --lts`
+   - Windows: `nvm install lts`
```
Comment (mikeharder) [inline CONTRIBUTING.md:65]: Recommend installing `lts` which should always be a good default for dev machines, and stays current as versions release over time.

**Review by konrad-jamrozik — APPROVED** [review]

**Comment by konrad-jamrozik** [comment]
Re publishing: I am ok either way. Let's wait for now?

Also:
- https://github.com/Azure/azure-openapi-validator/pull/669/files#r1515048617

**Pull request merged**